portal/src/graphql/adminapi/RolesScreen.tsx

```diff
@@ -64,36 +105,92 @@ const RolesScreen: React.VFC = function RolesScreen() {
     }
     return encodeOffsetToCursor(offset - 1);
   }, [isSearch, offset]);
-  const { data, loading } = useQuery<
+
+  const onChangeOffset = useCallback((offset) => {
+    setOffset(offset);
+  }, []);
+
+  const onChangeSearchKeyword = useCallback((_e, value) => {
+    if (value != null) {
+      setSearchKeyword(value);
+      // Reset offset when search keyword was changed.
+      setOffset(0);
+    }
+  }, []);
+
+  const onClearSearchKeyword = useCallback((_e) => {
+    setSearchKeyword("");
+    // Reset offset when search keyword was changed.
+    setOffset(0);
+  }, []);
+
+  const localSearchBoxProps: LocalSearchBoxProps = useMemo(() => {
+    return {
+      className: styles.searchBox,
+      placeholder: renderToString("search"),
+      value: searchKeyword,
+      onChange: onChangeSearchKeyword,
+      onClear: onClearSearchKeyword,
+    };
+  }, [
+    renderToString,
+    searchKeyword,
+    onChangeSearchKeyword,
+    onClearSearchKeyword,
+  ]);
+
+  const { data, loading, previousData } = useQuery<
     RolesListQueryQuery,
     RolesListQueryQueryVariables
   >(RolesListQueryDocument, {
     variables: {
-      searchKeyword,
-      pageSize,
+      pageSize: isSearch ? searchResultSize : pageSize,
+      searchKeyword: debouncedSearchKey,
       cursor,
     },
     fetchPolicy: "network-only",
   });
 
+  const isInitialLoading = loading && previousData == null;
+
   const items = useMemo(() => {
     return [{ to: ".", label: <FormattedMessage id="RolesScreen.title" /> }];
   }, []);
 
-  const isEmpty = !loading && data?.roles?.edges?.length === 0;
+  const isEmpty = !isInitialLoading && (data?.roles?.edges?.length ?? 0) === 0;
 
   return (
-    <ScreenContent className={styles.content} layout="list">
-      <div className={styles.widget}>
-        <NavBreadcrumb className="block" items={items} />
-      </div>
-      {isEmpty ? (
-        <RolesScreenEmptyState className={styles.widget} />
-      ) : (
-        // TODO: implement list view
-        <RolesScreenEmptyState className={styles.widget} />
-      )}
-    </ScreenContent>
+    <LocalSearchBoxContext.Provider value={localSearchBoxProps}>
+      <ScreenContent className={styles.content} layout="list">
+        <div className={styles.widget}>
+          <div className={styles.titleContainer}>
+            <NavBreadcrumb className="block" items={items} />
+            {!isEmpty ? <CreateRoleButton /> : null}
+          </div>
+          <LocalSearchBox />
+        </div>
+        {isEmpty ? (
+          isSearch ? (
+            <MessageBar className={cn(styles.widget, styles.message)}>
+              <FormattedMessage id="RolesScreen.empty.search" />
+            </MessageBar>
```

Comment:
Where did you find this design? IMO it should be displayed in the detail list.

Reply:
Thx. Missed the below design so followed the old design of users list

Reply:
Not so sure how to handle the empty state in mobile view? :sosad:

Expect it to be static or scroll with columns?

https://github.com/authgear/authgear-server/assets/63200569/f2370bc9-ab27-4ca8-967e-01eed44a51c1

Reply:
If scroll with columns, there is a chance that user cannot see it.
But static looks a bit weird too. Maybe @fungc-io What do you think?